### ISSUE TITLE
[Encoding] For certain files characters cross over the buffer boundaries

### ISSUE THREAD

**Issue by alphaCTzo7G** [vim/vim#3237]
This may be related to how `VIM` processes certain characters, even if they are valid `utf-8` characters: https://github.com/vim/vim/issues/3213

If I try to open the file: `https://github.com/pallets/jinja/blob/master/jinja2/_identifier.py` by cloning the repository to my /tst directory

the buffer contents are not in the same pane but cross over.. 

Note: I have `set encoding=utf-8` and even though `_idenfier.py` seems to contain garbled characters.. they are all valid `utf-8` characters. This is shown by `:e ++enc=utf-8`, `8g8` or `grep -axv '.*' _identifier.py`

![2018-07-20_11-34-19](https://user-images.githubusercontent.com/29691914/43019470-f340d78a-8c11-11e8-8e4f-3abe7385f7b8.gif)

I am using `tmux` over ssh.. but I have also used it in the GNOME terminal without `tmux` running and the same problem can be reproduced.

![2018-07-20_11-44-27](https://user-images.githubusercontent.com/29691914/43019581-5c12bb7a-8c12-11e8-80f3-ee6f29ffba8b.gif)

Is this supposed to happen? Why so?

**Comment by k-takata** [comment]
Isn't the first screenshot a problem of tmux?
How about just doing `cat _identifier.py`?

> but I have also used it in the GNOME terminal without tmux running and the same problem can be reproduced.

They might be different problems.

**Comment by chrisbra** [comment]
I think this happens, because Vim and the underlying terminal libraries have different opions on the utf8 properties.

**Comment by alphaCTzo7G** [comment]
@k-takata.. thanks for the suggestion.

While run `tmux` remotely over ssh, have multiple splits in `tmux`, and if I do `cat _identifier.py`, the same thing happens:
![2018-07-21_09-03-31](https://user-images.githubusercontent.com/29691914/43037699-30a2ac94-8cc5-11e8-9f15-f8d6bd4f7d40.gif)

In this case, `echo $TERM` shows `screen-256color`..

**Comment by alphaCTzo7G** [comment]
If I run `cat _identifier.py` directly on the `gnome terminal` while logged in locally and running `tmux`, it is able to correctly contain the characters within the panes

![virtualbox_2018-07-21_09-32-23](https://user-images.githubusercontent.com/29691914/43038166-7a22e2ae-8cc9-11e8-90cc-5481dd2fb6ae.png)

`echo $TERM` still gives `screen-256color`..

From the post before.. if I try to open `_identifier.py` in `vim` in  `gnome-terminal` without running `tmux`, I can reproduce the problem of overflowing characters across `vim-pane` boundaries:
![vim_gnome_terminal](https://user-images.githubusercontent.com/29691914/43019581-5c12bb7a-8c12-11e8-80f3-ee6f29ffba8b.gif)

So the problem seems to happen both in `vim` and for some reason when `tmux` is run over ssh.. but probably has differing root causes as @k-takata pointed out. But doesn't seem to be directly related to `tmux` itself.

How can this be solved?

**Comment by alphaCTzo7G** [comment]
@chrisbra.. is there.. is there a way to solve what you mentioned and reduce the differing opinions? Are you guys able to reproduce the problem?

**Comment by lifepillar** [comment]
@alphaCTzo7G Re your first screenshot, I can reproduce it in macOS with Terminal.app, but not with iTerm2 or MacVim. I had a similar issue (https://github.com/vim/vim/issues/2599) and the conclusion there was that the terminal or multiplexer is at fault, not Vim.

Your issue looks strikingly similar. For instance, if you type `<c-v>u002d<c-v>u036f` while in Insert mode in a new vertical split, you will see the rendering defect (btw, if you set `cc` or `cuc`, you will notice that `colorcolumn` and `cursorcolumn` are affected, too). U002D+U036F is one of the glyphs in `_identifier.py` (at line 2, column 13). To get information about any of those glyphs, put the cursor over any of them and type `ga`.

**Comment by alphaCTzo7G** [comment]
@lifepillar .. thanks.. I posted this in the `tmux` community as well. 

I am on Ubuntu.. so I tried `urxvt` as well. as its suppoed to have unicode support.. while `cat _identifier.py` doesn't overflow `tmux` pane-boundaries in `urxvt`.. 
![xming_2018-07-21_12-58-08](https://user-images.githubusercontent.com/29691914/43039624-fe348c2a-8ce5-11e8-9924-d00aa7a4f9ec.png)

If I run `vim` in `urxvt`, without running `tmux`, the same problem can be reproduced.. 

![xming_2018-07-21_12-58-58](https://user-images.githubusercontent.com/29691914/43039626-0c0e2e14-8ce6-11e8-9aef-aaccc6f459de.png)

So it seems that If its a terminal issue, shouldn't `cat _identifier.py` also result in overflow across the `tmux-pane` boundaries?

When I run `urxvt` why is the overflow happening only if I run `vim` in `urxvt/Gnome terminal`, but not when I run `tmux` in `urxvt/Gnome Terminal`?
